contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -2554,3 +2558,557 @@ EOMONTH(PG_FUNCTION_ARGS)
+
+	if (PG_ARGISNULL(0) || PG_ARGISNULL(1))
+		PG_RETURN_NULL();
+	else
+	{
+		object_id = (Oid) PG_GETARG_INT32(0);
+		property = lowerstr(text_to_cstring(PG_GETARG_TEXT_P(1)));
+		remove_trailing_spaces(property);
+	}
+
+	/*
+	 * Search for the object_id in pg_class, pg_proc, pg_attrdef, pg_constraint.
+	 * If the object_id is not found in any of the above catalogs, return NULL.
+	 * Else, get the object name, type of the object and the schema_id in which 
+	 * the object is present.
+	 */
+
+	/* pg_class */
+	tuple = SearchSysCache1(RELOID, ObjectIdGetDatum(object_id));
+	if (HeapTupleIsValid(tuple))
+	{
+		Form_pg_class pg_class = (Form_pg_class) GETSTRUCT(tuple);
+
+		found = true;
+		object_name = NameStr(pg_class->relname);
+
+		if (pg_class_aclcheck(object_id, user_id, ACL_SELECT) == ACLCHECK_OK)
+			schema_id = get_rel_namespace(object_id);
+
+		/* 
+		 * Get the type of the object 
+		 */
+		if ((pg_class->relpersistence == 'p' || pg_class->relpersistence == 'u' || pg_class->relpersistence == 't') &&
+				(pg_class->relkind == 'r'))
+		{
+			/* 
+			 * Check whether it is a Table type (TT) object.
+			 * The reltype of the pg_class object should be there in pg_type. The pg_type object found
+			 * should be of composite type (c) and the type of dependency should be DEPENDENCY_INTERNAL (i).
+			 * We scan pg_depend catalog to find the type of the dependency.
+			 */
+			HeapTuple tp;
+			tp = SearchSysCache1(TYPEOID, ObjectIdGetDatum(pg_class->reltype));
+			if(HeapTupleIsValid(tp))
+			{
+				Form_pg_type typform = (Form_pg_type) GETSTRUCT(tp);
+
+				if (typform->typtype == 'c')
+				{
+					Relation	depRel;
+					ScanKeyData key[2];
+					SysScanDesc scan;
+					HeapTuple	tup;
+
+					depRel = table_open(DependRelationId, RowExclusiveLock);
+
+					ScanKeyInit(&key[0],
+								Anum_pg_depend_objid,
+								BTEqualStrategyNumber, F_OIDEQ,
+								ObjectIdGetDatum(typform->typrelid));
+					ScanKeyInit(&key[1],
+								Anum_pg_depend_refobjid,
+								BTEqualStrategyNumber, F_OIDEQ,
+								ObjectIdGetDatum(typform->oid));
+
+					scan = systable_beginscan(depRel, InvalidOid, false,
+							  				NULL, 2, key);
+
+					if (HeapTupleIsValid(tup = systable_getnext(scan)))
+					{
+						Form_pg_depend depform = (Form_pg_depend) GETSTRUCT(tup);
+
+						if (depform->deptype == 'i')
+							type = "TT";
+					}
+
+					systable_endscan(scan);
+
+					table_close(depRel, RowExclusiveLock);
+				}
+			}
+			/*
+			 * If the object is not of Table type (TT), it should be user defined table (U)
+			 */
+			if (type == NULL || pg_strcasecmp(type, "TT") != 0)
+				type = "U";
+		}
+		else if (pg_class->relkind == 'v')
+			type = "V";
+		else if (pg_class->relkind == 's')
+			type = "SO";
+	}
+	/* pg_proc */
+	if (!found)
+	{
+		tuple = SearchSysCache1(PROCOID, ObjectIdGetDatum(object_id));
+		if (HeapTupleIsValid(tuple))
+		{
+			found = true;
+			if (pg_proc_aclcheck(object_id, user_id, ACL_EXECUTE) == ACLCHECK_OK)
```

Comment:
What is the behaviour for sql server when user does not have permission for that object?
Lets remember to add test case for that